### PR TITLE
Enhance Buffer Data Processing Pipeline

### PR DESCRIPTION
**Pull request by pan-x-c**
## Description

Implementation of #158 

- [x] **Add `ExperienceOperator` interface**, allowing users to implement this interface to process experience data in a pipeline manner.
- [x] **Add `ExperiencePipeline` actor**, responsible for continuously running operators.
- [x] **Merge `AddStrategy` into Operator** to avoid functional overlap.
- [x] **Optimize Data-Juicer service call method** to avoid dependency conflicts.
- [x] **Add `TaskPipeline`**, responsible for pre-processing the raw task datasets before exploring. (Only support using Data-Juicer operators for now)
- [x] **Add Operator developer guideline** to guide users on how to write their own Operators.

image

## Checklist

Please check the following items before code is ready to be reviewed.

- [x]  Code has passed all tests
- [x]  Docstrings have been added/updated in Google Style
- [x]  Documentation has been updated
- [x]  Code is ready for review

### CODE REVIEW

File: trinity/algorithm/advantage_fn/grpo_advantage.py [hunk 1]
```diff
@@ -81,3 +89,118 @@ def default_args(cls) -> Dict:
...
+    def group_experiences(self, exps):
+        return group_by(exps, id_type="task")
+
+    def calculate_group_advantage(
+        self, group_id: str, exps: List[Experience]
+    ) -> Tuple[List[Experience], Dict]:
+        metrics = {}
+        with torch.no_grad():
+            if len(exps) == 1:
+                group_reward_mean = torch.tensor(0.0)
```
Comment (yanxi-chen) [inline trinity/algorithm/advantage_fn/grpo_advantage.py]: I've been wondering for a while whether it would be better to set `group_reward_mean = exps[0].reward` when `len(exps) == 1`, so that this single experience has advantage 0, which better matches the rationale of GRPO. 

If a single experience is kept, its policy gradient may be still unbiased (in theory, for on-policy setting), but possibly with high variance (since this falls back to vanilla REINFORCE without baseline). 

Also, setting `group_reward_mean = torch.tensor(0.0)` can cause a bias in the output metric.

Maybe let's keep it as is for now, and investigate in the future if this makes any difference.

File: trinity/algorithm/advantage_fn/grpo_advantage.py [hunk 2]
```diff
@@ -81,3 +89,118 @@ def default_args(cls) -> Dict:
...
+            return []
+        else:
+            idx = torch.randint(0, len(valid_groups), (skipped_group_num,)).tolist()
+            duplicated_groups = [copy.deepcopy(valid_groups[i]) for i in idx]
+            duplicated_exps = [exp for group in duplicated_groups for exp in group]
+            exps = [exp for group in valid_groups for exp in group]
+            for exp in duplicated_exps:
+                exp.eid.task += (
+                    original_group_num  # Ensure unique task IDs for duplicated experiences
+                )
```
Comment (yanxi-chen) [inline trinity/algorithm/advantage_fn/grpo_advantage.py]: Just curious, it seems to me that
(1) modifying `exp.eid.task` is unnecessary here? group advantages have been already calculated, and task id might not get used anymore.
(2) this line might not truly "ensure unique task IDs for duplicated experiences" like the comment says? there might exist exps with the same task id in `duplicated_exps`, and they remain the same after adding the same `original_group_num`.

My bet is that these lines can be safely removed.

File: trinity/common/config.py [hunk 3]
```diff
@@ -239,8 +320,8 @@ class AlgorithmConfig:
     repeat_times: int = 1
 
     # the strategy for adding experiences to the buffer
-    add_strategy: Optional[str] = None
-    add_strategy_args: Optional[dict] = None
+    add_strategy: Optional[str] = None  # deprecated
+    add_strategy_args: Optional[dict] = None  # deprecated
```
Comment (yanxi-chen) [inline trinity/common/config.py]: Should these be removed completely, rather than merely "deprecated", since they no longer take effect?

**Review by yanxi-chen — COMMENTED** [review]
Some minor comments

**Review by yanxi-chen — APPROVED** [review]
lgtm
